lib/widgets/action_sheet.dart

```diff
@@ -1,36 +1,182 @@
+  MessageActionSheetMenuItemButton({
+    super.key,
+    required this.message,
+    required this.bottomSheetContext,
+    required this.messageListContext,
+  }) : assert(bottomSheetContext.findAncestorWidgetOfExactType<BottomSheet>() != null),
+       assert(messageListContext.findAncestorWidgetOfExactType<MessageListPage>() != null);
+
+  IconData get icon;
+  String get label;
+  VoidCallback get onPressed;
+
+  final Message message;
+  final BuildContext bottomSheetContext;
+  final BuildContext messageListContext;
+
+  @override
+  Widget build(BuildContext context) {
+    return MenuItemButton(
+      leadingIcon: Icon(icon),
+      onPressed: onPressed,
+      child: Text(label));
+  }
+}
+
+class ShareButton extends MessageActionSheetMenuItemButton {
+  ShareButton({
+    super.key,
+    required super.message,
+    required super.bottomSheetContext,
+    required super.messageListContext,
+  });
+
+  @override IconData get icon => Icons.adaptive.share;
+
+  @override String get label => 'Share';
+
+  @override VoidCallback get onPressed => () async {
+    // Close the message action sheet; we're about to show the share
+    // sheet. (We could do this after the sharing Future settles, but
+    // on iOS I get impatient with how slowly our action sheet
+    // dismisses in that case.)
+    // TODO(#24): Fix iOS bug where this call causes the keyboard to
+    //   reopen (if it was open at the time of this
+    //   `showMessageActionSheet` call) and cover a large part of the
+    //   share sheet.
+    Navigator.of(bottomSheetContext).pop();
+
+    // TODO: to support iPads, we're asked to give a
+    //   `sharePositionOrigin` param, or risk crashing / hanging:
+    //     https://pub.dev/packages/share_plus#ipad
+    //   Perhaps a wart in the API; discussion:
+    //     https://github.com/zulip/zulip-flutter/pull/12#discussion_r1130146231
+    // TODO: Share raw Markdown, not HTML
+    await Share.shareWithResult(message.content);
+  };
+}
+
+class QuoteAndReplyButton extends MessageActionSheetMenuItemButton {
+  QuoteAndReplyButton({
+    super.key,
+    required super.message,
+    required super.bottomSheetContext,
+    required super.messageListContext,
+  });
+
+  @override IconData get icon => Icons.format_quote_outlined;
+
+  @override String get label => 'Quote and reply';
+
+  @override VoidCallback get onPressed => () async {
+    // Close the message action sheet. We'll show the request progress
+    // in the compose-box content input with a "[Quoting…]" placeholder.
+    Navigator.of(bottomSheetContext).pop();
+
+    // This will be null only if the compose box disappeared after the
+    // message action sheet opened, and before "Quote and reply" was pressed.
+    // Currently a compose box can't ever disappear, so this is impossible.
+    ComposeBoxController? composeBoxController = MessageListPageState.composeBoxControllerOf(messageListContext);
```

Comment:
nit:
```suggestion
    ComposeBoxController? composeBoxController =
      MessageListPageState.composeBoxControllerOf(messageListContext);
```

I know we've talked about avoiding "cliffhanger" newlines following Flutter-repo style, but I think this line is too long for that.

Reply:
Plus as a bonus:
```suggestion
    ComposeBoxController composeBoxController = MessageListPageState.composeBoxControllerOf(messageListContext)!;
```
when the line is a reasonable length, it becomes reasonable to put an important one-character bit of syntax at the end of it. And this seems like a more on-point place for that `!` than the next line.